feat(ngFor): add custom trackBy function support

kara opened the pull request:
Make it possible to track items in iterables in custom ways (e.g. by ID or index), rather than simply by identity.

Usage:

With template element:

```
@Component(
  template: `
   <template ngFor #item [ngForOf]="items" [ngForTrackBy]=customTrackBy">
      {{item}}
   </template>
   `
)
class MyComponent {
  customTrackBy(index: number, obj: any): any {
    return index;
  }
}
```

With *ngFor:

```
@Component(
  template: `
   <div *ngFor="#item of items; trackBy:customTrackBy">
      {{item}}
   </div>
   `
)
class MyComponent {
  customTrackBy(index: number, obj: any): any {
    return index;
  }
}
```

@mhevery  Did you want to review this?

Closes #4402

vsavkin commented on modules/angular2/src/core/change_detection/differs/default_iterable_differ.ts:
```diff
@@ -42,6 +43,10 @@ export class DefaultIterableDiffer implements IterableDiffer {
   private _removalsHead: CollectionChangeRecord = null;
   private _removalsTail: CollectionChangeRecord = null;
 
+  constructor(private _trackByFn?: TrackByFn) {
```
Can you do `_trackByFn: TrackByFn = trackByIdentity` instead? Try to do this if ts2dart supports it. If not, what you have is good.

kara replied:
Yeah, that's what I wanted to do too :)  It doesn't seem to be supported.

vsavkin commented on modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts:
```diff
@@ -324,5 +324,100 @@ export function main() {
+
+      });
+
+      it('should track duplicate reinsertion normally with track by function', () => {
```
I am confused about this case. Could you explain why we need to support this?

kara replied:
This is something we were already supporting for regular track-by-identity (with the `verifyReinsertion` code) to ensure that duplicates don't get re-ordered in unexpected ways.  I thought it should probably work the same way even when tracking by ID, so added a test.

vsavkin commented on modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts:
```diff
@@ -324,5 +324,100 @@ export function main() {
+
+      var createItem =
+          (val: string) => {
+            var item = StringMapWrapper.create();
```
You don't have to use StringMapWrapper. You could just use

```
var item = {'id': val};
``

kara replied:
I was getting a weird Dart error, so I ended up switching to StringMapWrapper.  I think the problem might have been that I was omitting the quotation marks around ID. Changing!

kara committed:
feat(ngFor): add custom trackBy function support

    Make it possible to track items in iterables in custom ways (e.g. by ID or index), rather than simply by identity.

mary-poppins commented:
Merging PR #6779 on behalf of @alexeagle to branch [presubmit-alexeagle-pr-6779](https://github.com/angular/angular/tree/presubmit-alexeagle-pr-6779).

johnpapa commented:
I see this unit test, but I cannot seem to get ngForTrackBy to work for my sample.

https://github.com/angular/angular/blob/cee2318110eeea115e5f6fc5bfc814cbaa7d90d8/modules/angular2/test/common/directives/ng_for_spec.ts#L364-L402

I added `[ngForTrackBy]="customTrackBy"` to my element in the template. Then I added the `customTrackBy` function the way the unit test shows it. But I am getting this error:

```
EXCEPTION: Template parse errors:
Can't bind to 'ngForTrackBy' since it isn't a known native property ("    <div class="mdl-cell mdl-cell--3-col"
        *ngFor="#character of characters | async"
        [ERROR ->][ngForTrackBy]="trackBy"
        (click)="gotoDetail(character)">

"): DashboardComponent@6:8
```

It seems it wants me to declare `ngForTrackBy` to the component, but I do not see where it is exported from any module by Angular. Nor is the unit test I referenced doing that.

What else is needed?

cc // @kara @wardbell
